feat: add clickbench benchmark

AdamGS opened the pull request:
Adds a [clickbench](https://github.com/ClickHouse/ClickBench) run to our datafusion benchmarks. ~Some queries fail, see #1326.~
Running this is pretty resource intensive as the dataset is ~100m rows/15GB of parquet.

AdamGS committed:
more benchmarks

AdamGS committed:
Fix lint

AdamGS committed:
.

robert3005 commented on bench-vortex/src/tpch/execute.rs:
```diff
@@ -44,9 +43,11 @@ pub async fn run_tpch_query(
-    let physical_plan = state.create_physical_plan(&optimized).await?;
-    let result = collect(physical_plan.clone(), state.task_ctx()).await?;
-    Ok(result)
+    plan.collect().await
```
I thought this doesn't run all the optimizations steps

AdamGS replied:
yeah, commented it out when I was debugging some issue, I'll add it back in a second

AdamGS committed:
enable plan optimization

AdamGS commented:
some benchmarks are failing, I'll fix them tomorrow

AdamGS committed:
seems to run now

AdamGS committed:
fix clippy

lwwmanning commented on bench-vortex/src/clickbench.rs:
```diff
@@ -0,0 +1,264 @@
+    let vortex_dir = file.parent().unwrap().join(if enable_compression {
+        "vortex_compressed"
+    } else {
+        "vortex_uncompressed"
```
No real point in uncompressed file?

AdamGS replied:
definitely some more cleanup to do here, ironing out some bugs first.

AdamGS committed:
remove code that was split into a different PR

AdamGS committed:
Canonicalize constant array with extension dtype (#1322)

also `storage_array` -> `storage_data` seems appropriate.

---------

Co-authored-by: Robert Kruszewski <github@robertk.io>

AdamGS committed:
Remove buggy queries

AdamGS committed:
more work

robert3005 committed:
all

robert3005 committed:
new

danking committed:
two digit queries (#1388)

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into adamg/clickbench

AdamGS committed:
.

lwwmanning commented on .github/workflows/ci.yml:
```diff
@@ -150,7 +150,7 @@ jobs:
 
   bench-test:
     name: "bench test"
-    runs-on: ubuntu-latest
```
does this make bench test materially faster?

AdamGS replied:
actually not necessary, I'll revert it.

lwwmanning commented on bench-vortex/benches/clickbench.rs:
```diff
@@ -0,0 +1,77 @@
+    let output_path = basepath.join("processed.parquet");
+
+    let final_parquet_path = idempotent(&output_path, |output_path| {
+        println!("Fixing parquet file");
```
lol wat...? what's the story here?

AdamGS replied:
The parquet file clickbench provides misses some more "complex" types, so we need to either cast everything on query time, or re-process it to have the correct timestamp columns. See clickhouse/clickbench#7.

lwwmanning replied:
cool. would be good to put a comment with a link to that issue

AdamGS committed:
revert change

AdamGS committed:
Add comment

AdamGS committed:
add comment

AdamGS commented:
Running clickbench OOMs our dedicated runners, I'll re-generate the file locally tomorrow and put it in some accessible location for future runs.

gatesn commented:
@AdamGS can we make this a bin/clickbench like we did for TPCH?

danking commented:
@AdamGS fwiw, I put the two parquet files in s3://vortex-bench-dev/clickbench/

danking committed:
Merge branch 'develop' into adamg/clickbench

danking commented:
🚢

AdamGS commented:
I want to pull the data from AWS first, and then I'll merge

AdamGS committed:
count on aws

AdamGS committed:
only download data for clickbench

github-actions[bot] committed:
chore: release v0.20.0 (#1333)

## 🤖 New release
* `vortex`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-alp`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-array`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-buffer`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-datetime-dtype`: 0.19.0 -> 0.20.0
* `vortex-dtype`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-error`: 0.19.0 -> 0.20.0
* `vortex-flatbuffers`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-proto`: 0.19.0 -> 0.20.0
* `vortex-scalar`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-fastlanes`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-bytebool`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-datetime-parts`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-dict`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-expr`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-file`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-io`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-ipc`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-sampling-compressor`: 0.19.0 -> 0.20.0 (⚠️ API breaking
changes)
* `vortex-fsst`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-roaring`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-runend`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-runend-bool`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)
* `vortex-zigzag`: 0.19.0 -> 0.20.0 (✓ API compatible changes)
* `vortex-datafusion`: 0.19.0 -> 0.20.0 (⚠️ API breaking changes)

### ⚠️ `vortex-array` breaking changes

```
--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_missing.ron

Failed in:
  function vortex_array::compute::unary::scalar_at_unchecked, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/unary/scalar_at.rs:30
  function vortex_array::compute::search_sorted_u64, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/search_sorted.rs:161
  function vortex_array::array::compute_stats, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/varbin/stats.rs:22
  function vortex_array::elementwise::dyn_cast_array_iter, previously in file /tmp/.tmpcC0GjL/vortex-array/src/elementwise.rs:23
  function vortex_array::compute::search_sorted_u64_many, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/search_sorted.rs:207
  function vortex_array::compress::compute_pruning_stats, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compress.rs:72

--- failure function_parameter_count_changed: pub fn parameter count changed ---

Description:
A publicly-visible function now takes a different number of parameters.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#fn-change-arity
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_parameter_count_changed.ron

Failed in:
  vortex_array::compute::take now takes 3 parameters instead of 2, in /tmp/.tmprcY0ZS/vortex/vortex-array/src/compute/take.rs:43

--- failure inherent_method_missing: pub method removed or renamed ---

Description:
A publicly-visible method or associated fn is no longer available under its prior name. It may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/inherent_method_missing.ron

Failed in:
  StatsSet::new, previously in file /tmp/.tmpcC0GjL/vortex-array/src/stats/statsset.rs:22
  StatsSet::bools_with_true_count, previously in file /tmp/.tmpcC0GjL/vortex-array/src/stats/statsset.rs:98
  ArrayData::as_primitive, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/primitive/mod.rs:338
  BoolArray::from_vec, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/bool/mod.rs:123

--- failure method_parameter_count_changed: pub method parameter count changed ---

Description:
A publicly-visible method now takes a different number of parameters.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#fn-change-arity
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/method_parameter_count_changed.ron

Failed in:
  vortex_array::validity::Validity::take now takes 3 parameters instead of 2, in /tmp/.tmprcY0ZS/vortex/vortex-array/src/validity.rs:199

--- failure module_missing: pub module removed or renamed ---

Description:
A publicly-visible module cannot be imported by its prior path. A `pub use` may have been removed, or the module may have been renamed, removed, or made non-public.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/module_missing.ron

Failed in:
  mod vortex_array::array::visitor, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/visitor.rs:1
  mod vortex_array::elementwise, previously in file /tmp/.tmpcC0GjL/vortex-array/src/elementwise.rs:1

--- failure struct_missing: pub struct removed or renamed ---

Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/struct_missing.ron

Failed in:
  struct vortex_array::OwnedArrayData, previously in file /tmp/.tmpcC0GjL/vortex-array/src/data.rs:14
  struct vortex_array::TypedArray, previously in file /tmp/.tmpcC0GjL/vortex-array/src/typed.rs:15
  struct vortex_array::ViewedArrayData, previously in file /tmp/.tmpcC0GjL/vortex-array/src/view.rs:20

--- failure trait_added_supertrait: non-sealed trait added new supertraits ---

Description:
A non-sealed trait added one or more supertraits, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#generic-bounds-tighten
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_added_supertrait.ron

Failed in:
  trait vortex_array::ArrayTrait gained AsRef in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89
  trait vortex_array::ArrayTrait gained ArrayNBytes in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89

--- failure trait_method_added: pub trait method added ---

Description:
A non-sealed public trait added a new method without a default implementation, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#trait-new-item-no-default
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_method_added.ron

Failed in:
  trait method vortex_array::stats::Statistics::clear in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/stats/mod.rs:104
  trait method vortex_array::stats::Statistics::retain_only in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/stats/mod.rs:121

--- failure trait_method_missing: pub trait method removed or renamed ---

Description:
A trait method is no longer callable, and may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#major-any-change-to-trait-item-signatures
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_method_missing.ron

Failed in:
  method true_count of trait BoolArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:93
  method maybe_null_indices_iter of trait BoolArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:106
  method maybe_null_slices_iter of trait BoolArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:115
  method into_canonical of trait ArrayEncodingExt, previously in file /tmp/.tmpcC0GjL/vortex-array/src/encoding/mod.rs:80
  method scalar_at_unchecked of trait ScalarAtFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/unary/scalar_at.rs:9
  method search_sorted_u64 of trait SearchSortedFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/search_sorted.rs:103
  method search_sorted_u64_many of trait SearchSortedFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/search_sorted.rs:121
  method nbytes of trait ArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/lib.rs:318
  method u8_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:180
  method u16_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:184
  method u32_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:188
  method u64_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:192
  method i8_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:196
  method i16_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:200
  method i32_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:204
  method i64_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:208
  method f16_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:212
  method f32_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:216
  method f64_accessor of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:220
  method u8_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:224
  method u16_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:228
  method u32_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:232
  method u64_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:236
  method i8_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:240
  method i16_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:244
  method i32_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:248
  method i64_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:252
  method f16_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:256
  method f32_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:260
  method f64_iter of trait PrimitiveArrayTrait, previously in file /tmp/.tmpcC0GjL/vortex-array/src/variants.rs:264

--- failure trait_missing: pub trait removed or renamed ---

Description:
A publicly-visible trait cannot be imported by its prior path. A `pub use` may have been removed, or the trait itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_missing.ron

Failed in:
  trait vortex_array::elementwise::BinaryFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/elementwise.rs:7
  trait vortex_array::array::visitor::AcceptArrayVisitor, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/visitor.rs:7
  trait vortex_array::compute::ArrayCompute, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/mod.rs:31
  trait vortex_array::stats::ArrayStatisticsCompute, previously in file /tmp/.tmpcC0GjL/vortex-array/src/stats/mod.rs:115
  trait vortex_array::compute::MaybeCompareFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/compare.rs:77
  trait vortex_array::encoding::ArrayEncoding, previously in file /tmp/.tmpcC0GjL/vortex-array/src/encoding/mod.rs:50
  trait vortex_array::array::visitor::ArrayVisitor, previously in file /tmp/.tmpcC0GjL/vortex-array/src/array/visitor.rs:12
  trait vortex_array::compute::AndFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/boolean.rs:5
  trait vortex_array::elementwise::UnaryFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/elementwise.rs:15
  trait vortex_array::ToOwnedArrayData, previously in file /tmp/.tmpcC0GjL/vortex-array/src/lib.rs:299
  trait vortex_array::compute::OrFn, previously in file /tmp/.tmpcC0GjL/vortex-array/src/compute/boolean.rs:47

--- failure trait_removed_supertrait: supertrait removed or renamed ---

Description:
A supertrait was removed from a trait. Users of the trait can no longer assume it can also be used like its supertrait.
        ref: https://doc.rust-lang.org/reference/items/traits.html#supertraits
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_removed_supertrait.ron

Failed in:
  supertrait vortex_array::compute::ArrayCompute of trait ArrayTrait in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89
  supertrait vortex_array::array::visitor::AcceptArrayVisitor of trait ArrayTrait in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89
  supertrait vortex_array::stats::ArrayStatisticsCompute of trait ArrayTrait in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89
  supertrait vortex_array::ToOwnedArrayData of trait ArrayTrait in file /tmp/.tmprcY0ZS/vortex/vortex-array/src/lib.rs:89
```

### ⚠️ `vortex-dtype` breaking changes

```
--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_missing.ron

Failed in:
  function vortex_dtype::flatbuffers::deserialize_and_project, previously in file /tmp/.tmpcC0GjL/vortex-dtype/src/serde/flatbuffers/project.rs:27
```

### ⚠️ `vortex-flatbuffers` breaking changes

```
--- failure constructible_struct_adds_field: externally-constructible struct adds field ---

Description:
A pub struct constructible with a struct literal has a new pub field. Existing struct literals must be updated to include the new field.
        ref: https://doc.rust-lang.org/reference/expressions/struct-expr.html
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/constructible_struct_adds_field.ron

Failed in:
  field ArrayStatsArgs.uncompressed_size_in_bytes in /tmp/.tmprcY0ZS/vortex/vortex-flatbuffers/src/./generated/array.rs:442
```

### ⚠️ `vortex-file` breaking changes

```
--- failure auto_trait_impl_removed: auto trait no longer implemented ---

Description:
A public type has stopped implementing one or more auto traits. This can break downstream code that depends on the traits being implemented.
        ref: https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/auto_trait_impl_removed.ron

Failed in:
  type RowMask is no longer UnwindSafe, in /tmp/.tmprcY0ZS/vortex/vortex-file/src/read/mask.rs:19
  type RowMask is no longer RefUnwindSafe, in /tmp/.tmprcY0ZS/vortex/vortex-file/src/read/mask.rs:19

--- failure derive_trait_impl_removed: built-in derived trait no longer implemented ---

Description:
A public type has stopped deriving one or more traits. This can break downstream code that depends on those types implementing those traits.
        ref: https://doc.rust-lang.org/reference/attributes/derive.html#derive
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/derive_trait_impl_removed.ron

Failed in:
  type RowMask no longer derives PartialEq, in /tmp/.tmprcY0ZS/vortex/vortex-file/src/read/mask.rs:19
  type RowMask no longer derives Eq, in /tmp/.tmprcY0ZS/vortex/vortex-file/src/read/mask.rs:19

--- failure inherent_method_missing: pub method removed or renamed ---

Description:
A publicly-visible method or associated fn is no longer available under its prior name. It may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/inherent_method_missing.ron

Failed in:
  RowMask::and_inplace, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/mask.rs:178
  RowMask::to_mask_array, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/mask.rs:227
  InitialRead::fb_schema, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/builder/initial_read.rs:55
  VortexFileArrayStream::schema, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/stream.rs:78

--- failure pub_module_level_const_missing: pub module-level const is missing ---

Description:
A public const is missing, renamed, or changed from const to static.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/pub_module_level_const_missing.ron

Failed in:
  ALIGNMENT in file /tmp/.tmpcC0GjL/vortex-file/src/lib.rs:143

--- failure struct_missing: pub struct removed or renamed ---

Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/struct_missing.ron

Failed in:
  struct vortex_file::IoDispatcher, previously in file /tmp/.tmpcC0GjL/vortex-file/src/dispatcher/mod.rs:59
  struct vortex_file::layouts::ColumnarLayoutSpec, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/layouts/columnar.rs:22
  struct vortex_file::LazilyDeserializedDType, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/cache.rs:48
  struct vortex_file::layouts::InlineDTypeLayoutSpec, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/layouts/inline_dtype.rs:20
  struct vortex_file::Layout, previously in file /tmp/.tmpcC0GjL/vortex-file/src/write/layout.rs:11
  struct vortex_file::layouts::FlatLayoutSpec, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/layouts/flat.rs:19
  struct vortex_file::layouts::ChunkedLayoutSpec, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/layouts/chunked.rs:16

--- failure trait_missing: pub trait removed or renamed ---

Description:
A publicly-visible trait cannot be imported by its prior path. A `pub use` may have been removed, or the trait itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_missing.ron

Failed in:
  trait vortex_file::Dispatch, previously in file /tmp/.tmpcC0GjL/vortex-file/src/dispatcher/mod.rs:28
  trait vortex_file::LayoutSpec, previously in file /tmp/.tmpcC0GjL/vortex-file/src/read/context.rs:25
```

### ⚠️ `vortex-io` breaking changes

```
--- failure struct_missing: pub struct removed or renamed ---

Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/struct_missing.ron

Failed in:
  struct vortex_io::FuturesAdapter, previously in file /tmp/.tmpcC0GjL/vortex-io/src/futures.rs:8

--- failure trait_missing: pub trait removed or renamed ---

Description:
A publicly-visible trait cannot be imported by its prior path. A `pub use` may have been removed, or the trait itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_missing.ron

Failed in:
  trait vortex_io::VortexRead, previously in file /tmp/.tmpcC0GjL/vortex-io/src/read.rs:17
```

### ⚠️ `vortex-sampling-compressor` breaking changes

```
--- failure inherent_method_missing: pub method removed or renamed ---

Description:
A publicly-visible method or associated fn is no longer available under its prior name. It may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/inherent_method_missing.ron

Failed in:
  SamplingCompressor::compress_array, previously in file /tmp/.tmpcC0GjL/vortex-sampling-compressor/src/sampling_compressor.rs:161

--- failure pub_static_missing: pub static is missing ---

Description:
A public static is missing, renamed, or made private.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/pub_static_missing.ron

Failed in:
  FASTEST_COMPRESSORS in file /tmp/.tmpcC0GjL/vortex-sampling-compressor/src/lib.rs:54
```

### ⚠️ `vortex-runend` breaking changes

```
--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_missing.ron

Failed in:
  function vortex_runend::compress::runend_decode, previously in file /tmp/.tmpcC0GjL/vortex-runend/src/compress.rs:53

--- failure function_parameter_count_changed: pub fn parameter count changed ---

Description:
A publicly-visible function now takes a different number of parameters.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#fn-change-arity
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_parameter_count_changed.ron

Failed in:
  vortex_runend::compress::runend_decode_primitive now takes 5 parameters instead of 4, in /tmp/.tmprcY0ZS/vortex/encodings/runend/src/compress.rs:53
```

### ⚠️ `vortex-runend-bool` breaking changes

```
--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_missing.ron

Failed in:
  function vortex_runend_bool::compress::value_at_index, previously in file /tmp/.tmpcC0GjL/vortex-runend-bool/src/compress.rs:89
  function vortex_runend_bool::compress::runend_bool_encode, previously in file /tmp/.tmpcC0GjL/vortex-runend-bool/src/compress.rs:12
  function vortex_runend_bool::compress::runend_bool_decode, previously in file /tmp/.tmpcC0GjL/vortex-runend-bool/src/compress.rs:44

--- failure function_parameter_count_changed: pub fn parameter count changed ---

Description:
A publicly-visible function now takes a different number of parameters.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#fn-change-arity
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_parameter_count_changed.ron

Failed in:
  vortex_runend_bool::compress::runend_bool_decode_slice now takes 3 parameters instead of 4, in /tmp/.tmprcY0ZS/vortex/encodings/runend-bool/src/compress.rs:66
```

### ⚠️ `vortex-datafusion` breaking changes

```
--- failure method_parameter_count_changed: pub method parameter count changed ---

Description:
A publicly-visible method now takes a different number of parameters.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#fn-change-arity
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/method_parameter_count_changed.ron

Failed in:
  vortex_datafusion::persistent::execution::VortexExec::try_new now takes 4 parameters instead of 5, in /tmp/.tmprcY0ZS/vortex/vortex-datafusion/src/persistent/execution.rs:26

--- failure module_missing: pub module removed or renamed ---

Description:
A publicly-visible module cannot be imported by its prior path. A `pub use` may have been removed, or the module may have been renamed, removed, or made non-public.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/module_missing.ron

Failed in:
  mod vortex_datafusion::persistent::provider, previously in file /tmp/.tmpcC0GjL/vortex-datafusion/src/persistent/provider.rs:1

--- failure struct_missing: pub struct removed or renamed ---

Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/struct_missing.ron

Failed in:
  struct vortex_datafusion::persistent::provider::VortexFileTableProvider, previously in file /tmp/.tmpcC0GjL/vortex-datafusion/src/persistent/provider.rs:25
  struct vortex_datafusion::persistent::config::VortexTableOptions, previously in file /tmp/.tmpcC0GjL/vortex-datafusion/src/persistent/config.rs:36

--- failure trait_method_missing: pub trait method removed or renamed ---

Description:
A trait method is no longer callable, and may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#major-any-change-to-trait-item-signatures
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_method_missing.ron

Failed in:
  method register_disk_vortex_opts of trait SessionContextExt, previously in file /tmp/.tmpcC0GjL/vortex-datafusion/src/lib.rs:71
  method read_disk_vortex_opts of trait SessionContextExt, previously in file /tmp/.tmpcC0GjL/vortex-datafusion/src/lib.rs:78
```

<details><summary><i><b>Changelog</b></i></summary><p>


</p></details>

---
This PR was generated with
[release-plz](https://github.com/release-plz/release-plz/).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

a10y committed:
fix: regression for ObjectStoreReadAt (#1483)

Fixes a regression introduced by #1456 

See NOTE comment in code for explanation.

This change fixes the perf regression for me, at least for Q15 brings us
from 250ms back down to ~90ms


<img width="678" alt="image"
src="https://github.com/user-attachments/assets/2ca9b321-2a40-415d-8fba-2344a50c7794">

AdamGS committed:
feat: Layout metadata reader and column statistics (#1455)

Adds a dedicated file metadata reader and uses it to provide DataFusion
with file-level statistics.

---------

Co-authored-by: Will Manning <will@willmanning.io>
Co-authored-by: Daniel King <dan@spiraldb.com>

AdamGS committed:
permissions

AdamGS committed:
.

AdamGS committed:
set up dedicated role

AdamGS committed:
Cleanup all the clickbench leftovers

gatesn committed:
Flatten unary compute mod (#1489)

We probably ought to group functions at some point, e.g. by logical,
arithmetic, select, etc. But grouping by arity is weird...

gatesn committed:
Make BinaryBooleanFn consistent with CompareFn (#1488)

So now the implementation returns `Result<Option<ArrayData>>` and the
dispatch logic is almost identical.

AdamGS committed:
cleanup?

gatesn committed:
Remove uses of with_dyn for validity (#1487)

gatesn committed:
Move invert from BoolArrayTrait to InvertFn (#1490)

danking committed:
feat: prune layouts based on stats (#1485)

LayoutReaders can now check if a given row range is fully pruned by
their Scan expression (which must be Boolean-valued). Moreover,
PrunedSplitIterator uses this information to avoid even generating
RowMasks for ranges which do not contain any matching rows.

---------

Co-authored-by: Will Manning <will@willmanning.io>

renovate[bot] committed:
chore(deps): update rust crate tracing to v0.1.41 (#1493)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
| [tracing](https://tokio.rs)
([source](https://redirect.github.com/tokio-rs/tracing)) |
workspace.dependencies | patch | `0.1.40` -> `0.1.41` |

---

### Release Notes

<details>
<summary>tokio-rs/tracing (tracing)</summary>

###
[`v0.1.41`](https://redirect.github.com/tokio-rs/tracing/releases/tag/tracing-0.1.41):
tracing 0.1.41

[Compare
Source](https://redirect.github.com/tokio-rs/tracing/compare/tracing-0.1.40...tracing-0.1.41)

\[ [crates.io][crate-0.1.41] ] | \[ [docs.rs][docs-0.1.41] ]

This release updates the `tracing-core` dependency to
[v0.1.33][core-0.1.33] and
the `tracing-attributes` dependency to [v0.1.28][attrs-0.1.28].

##### Added

-   **core**: Add index API for `Field` ([#&#8203;2820])
- **core**: Allow `&[u8]` to be recorded as event/span field
([#&#8203;2954])

##### Changed

-   Bump MSRV to 1.63 ([#&#8203;2793])
-   **core**: Use const `thread_local`s when possible ([#&#8203;2838])

##### Fixed

-   Removed core imports in macros ([#&#8203;2762])
- **attributes**: Added missing RecordTypes for instrument
([#&#8203;2781])
- **attributes**: Change order of async and unsafe modifier
([#&#8203;2864])
-   Fix missing field prefixes ([#&#8203;2878])
-   **attributes**: Extract match scrutinee ([#&#8203;2880])
-   Fix non-simple macro usage without message ([#&#8203;2879])
- Fix event macros with constant field names in the first position
([#&#8203;2883])
-   Allow field path segments to be keywords ([#&#8203;2925])
-   **core**: Fix missed `register_callsite` error ([#&#8203;2938])
- **attributes**: Support const values for `target` and `name`
([#&#8203;2941])
- Prefix macro calls with ::core to avoid clashing with local macros
([#&#8203;3024])

[#&#8203;2762]: https://redirect.github.com/tokio-rs/tracing/pull/2762

[#&#8203;2781]: https://redirect.github.com/tokio-rs/tracing/pull/2781

[#&#8203;2793]: https://redirect.github.com/tokio-rs/tracing/pull/2793

[#&#8203;2820]: https://redirect.github.com/tokio-rs/tracing/pull/2820

[#&#8203;2838]: https://redirect.github.com/tokio-rs/tracing/pull/2838

[#&#8203;2864]: https://redirect.github.com/tokio-rs/tracing/pull/2864

[#&#8203;2878]: https://redirect.github.com/tokio-rs/tracing/pull/2878

[#&#8203;2879]: https://redirect.github.com/tokio-rs/tracing/pull/2879

[#&#8203;2880]: https://redirect.github.com/tokio-rs/tracing/pull/2880

[#&#8203;2883]: https://redirect.github.com/tokio-rs/tracing/pull/2883

[#&#8203;2925]: https://redirect.github.com/tokio-rs/tracing/pull/2925

[#&#8203;2938]: https://redirect.github.com/tokio-rs/tracing/pull/2938

[#&#8203;2941]: https://redirect.github.com/tokio-rs/tracing/pull/2941

[#&#8203;2954]: https://redirect.github.com/tokio-rs/tracing/pull/2954

[#&#8203;3024]: https://redirect.github.com/tokio-rs/tracing/pull/3024

[attrs-0.1.28]:
https://redirect.github.com/tokio-rs/tracing/releases/tag/tracing-attributes-0.1.28

[core-0.1.33]:
https://redirect.github.com/tokio-rs/tracing/releases/tag/tracing-core-0.1.33

[docs-0.1.41]: https://docs.rs/tracing/0.1.41/tracing/

[crate-0.1.41]: https://crates.io/crates/tracing/0.1.41

</details>

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Enabled.

♻ **Rebasing**: Whenever PR becomes conflicted, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/spiraldb/vortex).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiIzOS4xOS4wIiwidXBkYXRlZEluVmVyIjoiMzkuMTkuMCIsInRhcmdldEJyYW5jaCI6ImRldmVsb3AiLCJsYWJlbHMiOltdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore: release v0.21.0 (#1484)

## 🤖 New release
* `vortex`: 0.20.0 -> 0.21.0
* `vortex-alp`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-array`: 0.20.0 -> 0.21.0 (⚠️ API breaking changes)
* `vortex-buffer`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-datetime-dtype`: 0.20.0 -> 0.21.0
* `vortex-dtype`: 0.20.0 -> 0.21.0
* `vortex-error`: 0.20.0 -> 0.21.0
* `vortex-flatbuffers`: 0.20.0 -> 0.21.0
* `vortex-proto`: 0.20.0 -> 0.21.0
* `vortex-scalar`: 0.20.0 -> 0.21.0
* `vortex-fastlanes`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-bytebool`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-datetime-parts`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-dict`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-expr`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-file`: 0.20.0 -> 0.21.0 (⚠️ API breaking changes)
* `vortex-io`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-ipc`: 0.20.0 -> 0.21.0
* `vortex-sampling-compressor`: 0.20.0 -> 0.21.0
* `vortex-fsst`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-roaring`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-runend`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-runend-bool`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-zigzag`: 0.20.0 -> 0.21.0 (✓ API compatible changes)
* `vortex-datafusion`: 0.20.0 -> 0.21.0 (✓ API compatible changes)

### ⚠️ `vortex-array` breaking changes

```
--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/function_missing.ron

Failed in:
  function vortex_array::compute::unary::try_cast, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/cast.rs:30
  function vortex_array::compute::unary::scalar_at, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/scalar_at.rs:30
  function vortex_array::compute::unary::fill_forward, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/fill_forward.rs:31
  function vortex_array::compute::unary::subtract_scalar, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/scalar_subtract.rs:28

--- failure module_missing: pub module removed or renamed ---

Description:
A publicly-visible module cannot be imported by its prior path. A `pub use` may have been removed, or the module may have been renamed, removed, or made non-public.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/module_missing.ron

Failed in:
  mod vortex_array::compute::unary, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/mod.rs:1

--- failure trait_method_missing: pub trait method removed or renamed ---

Description:
A trait method is no longer callable, and may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#major-any-change-to-trait-item-signatures
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_method_missing.ron

Failed in:
  method invert of trait BoolArrayTrait, previously in file /tmp/.tmp8nwQKL/vortex-array/src/variants.rs:90

--- failure trait_missing: pub trait removed or renamed ---

Description:
A publicly-visible trait cannot be imported by its prior path. A `pub use` may have been removed, or the trait itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_missing.ron

Failed in:
  trait vortex_array::compute::unary::ScalarAtFn, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/scalar_at.rs:10
  trait vortex_array::compute::unary::CastFn, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/cast.rs:7
  trait vortex_array::compute::unary::SubtractScalarFn, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/scalar_subtract.rs:8
  trait vortex_array::compute::unary::FillForwardFn, previously in file /tmp/.tmp8nwQKL/vortex-array/src/compute/unary/fill_forward.rs:11
```

### ⚠️ `vortex-file` breaking changes

```
--- failure trait_method_added: pub trait method added ---

Description:
A non-sealed public trait added a new method without a default implementation, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#trait-new-item-no-default
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.36.0/src/lints/trait_method_added.ron

Failed in:
  trait method vortex_file::LayoutReader::read_metadata in file /tmp/.tmpe7MhwT/vortex/vortex-file/src/read/mod.rs:110
  trait method vortex_file::LayoutReader::can_prune in file /tmp/.tmpe7MhwT/vortex/vortex-file/src/read/mod.rs:113
```

<details><summary><i><b>Changelog</b></i></summary><p>


</p></details>

---
This PR was generated with
[release-plz](https://github.com/release-plz/release-plz/).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

danking committed:
fix: properly gate things by features & test for that (#1494)

Ideally we would test every combination of features, but at the very
least, we should verify that no features, default features, and all
features pass clippy.

The two main changes:

1. In vortex-io, we need a syntactically valid expressoin when neither
compio nor tokio are available.
2. In vortex-array, we need both flatbuffers and flexbuffers for views
to work. I removed the optional dependencies on these libraries. In
theory, someone could go through our code base and properly feature flag
array views, but that began to feel onerous as I did it.

AdamGS committed:
more things

AdamGS committed:
100 file variant

AdamGS committed:
fmt

AdamGS committed:
should work

AdamGS committed:
use provided schema

AdamGS committed:
Merge branch 'develop' into adamg/clickbench

AdamGS committed:
Merge branch 'develop' into adamg/clickbench

AdamGS committed:
Merge branch 'develop' into adamg/clickbench

AdamGS committed:
Install duckdb plugin

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
oops

AdamGS committed:
Set benchmark run name
